Converge common msbuild properties and targets

ViktorHofer opened the pull request:
Move common msbuild properties and targets into the repo root.

@trylek I don't really understand the difference between dir.common.props/targets and Directory.Build.props/targets. I think we should clean this up so that only the latter exists and is imported everywhere.

ViktorHofer committed:
Converge common msbuild properties and targets

Moving common msbuild properties and targets into the repo root.

ViktorHofer committed:
Update CoreLib tfm

jkotas commented on Directory.Build.props:
```diff
@@ -49,6 +57,49 @@
+  <!-- Analyzers -->
+  <PropertyGroup>
+    <CodeAnalysisRuleset>$(LibrariesProjectRoot)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <EnablePinvokeUWPAnalyzer>false</EnablePinvokeUWPAnalyzer>
```
We have dropped UWP target. Is this property still needed? And if it is still needed, is it relevant to anything outside libraries?

ViktorHofer replied:
This property affects every project that references `Microsoft.DotNet.CodeAnalysis`: https://github.com/dotnet/arcade/blob/7584330809c5463245e24f0f86ac2531a36dacd5/src/Microsoft.DotNet.CodeAnalysis/build/Microsoft.DotNet.CodeAnalysis.targets#L6. The right fix here is to flip the bit there. I will submit a PR.

ViktorHofer replied:
https://github.com/dotnet/arcade/pull/4446

jkotas commented on Directory.Build.props:
```diff
@@ -49,6 +57,49 @@
+    <EnablePinvokeUWPAnalyzer>false</EnablePinvokeUWPAnalyzer>
+  </PropertyGroup>
+
+  <!-- Set some shortcuts for more terse conditions in project files -->
```
Is this relevant to anything outside libraries? I would expect everything outside libraries targets netcoreapp.

ViktorHofer replied:
There is at least tools-local which multi-targets and I'm sure there are other cases. That said, I don't have a strong opinion.

ViktorHofer replied:
I had to move those back anyway as they rely on configurations set in libraries which happen at a later point.

jkotas commented on Directory.Build.targets:
```diff
@@ -1,5 +1,33 @@
+
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(SkipImportArcadeSdkFromRoot)' != 'true'" />
 
+  <!-- Adds Nullable annotation attributes to netstandard <= 2.0 builds -->
```
Is this relevant to anything outside libraries?

ViktorHofer replied:
Any library outside libraries that wants to enable nullable annotations which doesn't target .NET Core >= 3.0 benefits from this. I think there are cases in the other partitions where this might be useful.

jkotas replied:
What are those cases? I expect that we will move all libraries that ship as netstandard targeting NuGet packages to live under libraries.

ViktorHofer replied:
That's a good argument. Will move it back. I was thinking about enabling this scenario just entirely in the repo but if there's no use case for it then this might be over-engineering.

jkotas commented on src/libraries/Directory.Build.props:
```diff
@@ -119,16 +97,16 @@
   <PropertyGroup>
     <!-- By default make all libraries to be AnyCPU but individual projects can override it if they need to -->
     <Platform>AnyCPU</Platform>
-    <OutputType>Library</OutputType>
```
Just curious - where did this go?

ViktorHofer replied:
this is the default, no need to be explicit here.

jkotas commented on Directory.Build.props:
```diff
@@ -49,6 +57,49 @@
+
+  <!-- Analyzers -->
+  <PropertyGroup>
+    <CodeAnalysisRuleset>$(LibrariesProjectRoot)CodeAnalysis.ruleset</CodeAnalysisRuleset>
```
Majority of the C# code outside libraries is not polished to follow all these codeanalysis rules. Should this stay specific to libraries?

ViktorHofer replied:
That file is used by CoreLib, which lives outside libraries, as well. Extracting it into common makes most sense.

jkotas replied:
CoreLib is referencing all its source files from under libraries, so it can reference the codeanalysis ruleset from under libraries as well.

ViktorHofer replied:
> CoreLib is referencing all its source files from under libraries

https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj#L137-L412?

jkotas replied:
I meant to say`CoreLib is referencing a lot of of its source files from under libraries.

ViktorHofer replied:
If the plan is to move all managed projects to libraries than I agree with what you are saying but I don't think that's the plan? Also this property doesn't trigger thousands of errors/warnings but just sets the path to the rulset file in case a project references the analyzers. cc @stephentoub for dicussion.

jkotas replied:
We are going to have many managed projects outside libraries.

jkotas replied:
> just sets the path to the rulset file in case a project references the analyzers

Ok, if this is a link to file with reasonable set of defaults, this can be set globally. Should the file move to a more global place too then?

jkotas replied:
Also, what one has to do to enable use of the default ruleset? Should we also have the references to the analyzer packages, etc. here so that one can enable use of the default ruleset via a single property?

ViktorHofer replied:
> Also, what one has to do to enable use of the default ruleset? Should we also have the references to the analyzer packages, etc. here so that one can enable use of the default ruleset via a single property

AFAIK that's @stephentoub's plan.

> Should the file move to a more global place too then?

I proposed that as well but @stephentoub was against it. I think if we move all logic into common we should move that file as well.

stephentoub replied:
I wasn't against it. I just didn't want to expand that PR to consolidate all the logic that's currently spread across several places. I'm happy for all analysis support for managed code in the repo to be consolidated to one top-level location in the repo.

ViktorHofer replied:
I was referring to this comment: https://github.com/dotnet/runtime/pull/59#issuecomment-554401215. Anyways as you just agreed with moving everything analyzer related into root, I'm happy to do so :)

stephentoub replied:
Right. If we're only analyzing src in libraries (plus a few corelib files), it shouldn't be at the root. But this discussion suggests we're talking about enabling it for lots of stuff outside of libraries. Also, I think it should be consolidated, but not at the root of the repo. Isn't it better in eng?

ViktorHofer replied:
The [analyzers.props](https://github.com/dotnet/runtime/blob/master/eng/analyzers.props) file already lives in eng. I think it makes most sense to move the property that points to the ruleset file there too. We can move the rulset file there as well, I don't have a strong opinion on that.

stephentoub replied:
Sounds like the ruleset reference should just be added to that file and the ruleset put next to it.

ViktorHofer replied:
Perfect.

ViktorHofer committed:
Consolidate analyzers logic into the repo root

Consolidating msbuild analyzers logic centrally in the repo root and
adding a property switch `EnableAnalyzers` to enable analyzers. We
still need to import the Analyzers.props file unconditionally as
libraries currently doesn't use the PackageReference logic in ref and
src projects and instead uses a depproj to create a props file
dynamically.

ViktorHofer committed:
Fix build break

ViktorHofer committed:
PR feedback

trylek reviewed (approved):
Looks good, thanks for cleaning this up. I'll take a look at the extra CoreCLR scripts.

ViktorHofer commented:
Had to force push without changing anything (Arcade update wasn't picked up in a retry).

ViktorHofer committed:
Fix build in libraries

ViktorHofer committed:
Move import around as ruleset is needed for xunit

ViktorHofer committed:
Package import analyzers correctly

ViktorHofer committed:
More cleanup to fix CoreLib build

jkotas commented on src/coreclr/src/System.Private.CoreLib/ILLink.targets:
```diff
@@ -1,146 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
There is important logic in this file that auto-generates the ILLinkXml file based on the unmanaged runtime dependencies. It is not ok to just delete it.

If you want to refactor this or clean it up, it would be best done as separate PR.

ViktorHofer replied:
I diffed the contents and went through git history. It's a duplicate of eng/illink.targets.

ViktorHofer replied:
The special thing that you are referring to is probably https://github.com/dotnet/runtime/blob/master/src/coreclr/src/System.Private.CoreLib/CreateRuntimeRootILLinkDescriptorFile.targets which will stay as is.

ViktorHofer committed:
Un-condition illink restore

ViktorHofer committed:
Import condition fixes

jkotas commented on eng/illink.targets:
```diff
@@ -24,7 +24,7 @@
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(ILLinkTrimXml)' != ''">
+  <ItemGroup Condition="Exists('$(ILLinkTrimXml)')">
```
The exists condition like this just make for hard to debug problems. If one specifies ILLinkTrimXml file, it should better exist.

ViktorHofer replied:
reverted

jkotas commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -20,6 +18,10 @@
+
+    <ILLinkClearInitLocals>true</ILLinkClearInitLocals>
+    <ILLinkTrimAssembly>true</ILLinkTrimAssembly>
+    <ILLinkTrimXml>unused</ILLinkTrimXml>
```
Should this rather point to the auto-generated ILLinkTrimXml under obj directory?

ViktorHofer replied:
You are absolutely right. Pushed a commit.

jkotas commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -96,10 +98,6 @@
-    <AssemblyName>System.Private.CoreLib</AssemblyName>
-    <AssemblyVersion>5.0.0.0</AssemblyVersion>
-    <ExcludeAssemblyInfoPartialFile>true</ExcludeAssemblyInfoPartialFile>
-    <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute>
```
Why do we need to generate the target framework atrribute in CoreLib? It is just an extra cruft that nobody cares about.

ViktorHofer replied:
We generate the attribute for all assemblies in libraries, even for Private ones like System.Private.Xml. Mainly I did this to follow-up on your comment here: https://github.com/dotnet/corefx/pull/32680#issuecomment-429478625

jkotas replied:
Ah, ok - I forgot about that discussion. As you can tell, I still wonder why we are generating this cruft.

ViktorHofer replied:
> I still wonder why we are generating this cruft for inbox assemblies.

It's funny because in the past you mentioned that we should enable it for all - even the inbox ones. I don't have a strong opinion here but I would like to enable this for the sake of consistency. Also I prefer not opting out of the default sdk behavior if there isn't a strong reason.

Also as a side note, I noticed partners taking a dependency on the attribute, ie vstest for test assemblies: https://github.com/microsoft/vstest/blob/7b6248203164f8ea821f6795632bd22e0b69afb0/src/Microsoft.TestPlatform.ObjectModel/Utilities/AssemblyLoadWorker.cs#L67.

jkotas replied:
> It's funny because in the past you mentioned that we should enable it for all - even the inbox ones.

I do not think that I have said that. I [have said](https://github.com/dotnet/corefx/pull/32680#issuecomment-429478625) that if all inbox ones have it, CoreLib should have it too.

>  vstest for test assemblies:

These are not inbox assemblies.

ViktorHofer replied:
> I have said that if all inbox ones have it, CoreLib should have it too.

Thanks for clarifying.

> And I believe that this path is .NET Framework specific.

I know there was another code path somewhere for .NET Core but I can't find it right now...

ViktorHofer replied:
If you think we should remove the property from all inbox assemblies we can do that in a follow-up PR.

ViktorHofer committed:
Fix of Subsets cleanup

ViktorHofer committed:
ILLink feedback

ViktorHofer committed:
Allow different illink.tasks location for clr

ViktorHofer committed:
Simplify libraries illink infra

jkotas commented on src/coreclr/src/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -62,6 +63,7 @@
     <_FullFrameworkReferenceAssemblyPaths>$(MSBuildThisFileDirectory)/Documentation</_FullFrameworkReferenceAssemblyPaths>
     <SkipCommonResourcesIncludes>true</SkipCommonResourcesIncludes>
     <DocumentationFile>$(OutputPath)$(MSBuildProjectName).xml</DocumentationFile>
+    <EnableAnalyzers>false</EnableAnalyzers>
```
@ViktorHofer Why did you disable the analyzers for CoreLib? They were enabled before your change.

jkotas replied:
Opened https://github.com/dotnet/runtime/issues/422 to have this tracked

ViktorHofer replied:
https://github.com/dotnet/runtime/pull/424/files

safern reviewed (commented):
LGTM
